### PR TITLE
Append `aws-chunked` to `Content-Encoding` header if present

### PR DESCRIPTION
**Pull request by dlm6693**
Primarily in various S3 operations, users may provide a `ContentEncoding` parameter. This tells the service how the content that is being sent/requested is encoded. An example of this would be `gzip`. However, for streaming AWS operations (i.e. ones that have trailers) where a `ChecksumAlgorithm` parameter is provided, the SDK is currently overriding any user input to `aws-chunked`. Instead this special content encoding should be appended to the user input.

### CODE REVIEW

File: tests/unit/test_httpchecksum.py
Comment (nateprewitt) [inline tests/unit/test_httpchecksum.py:304]: We may want to either do a one-off manual test or add a functional test to ensure operations that bind input to the `Content-Encoding` header aren't affected when a checksum is used. S3 PutObject would be the best place to verify that.

File: tests/unit/test_httpchecksum.py
Comment (nateprewitt) [inline tests/unit/test_httpchecksum.py:315]: This may be more accurate. We are modifying it, but we don't want to overwrite it.
```suggestion
        # The content encoding should only have been appended
```

File: tests/unit/test_httpchecksum.py
Comment (dlm6693) [reply tests/unit/test_httpchecksum.py:304]: Oh sorry I didn't mention in the PR that I did some manual testing making actual `put_object` requests to a test bucket. I did think this may be worthy of a new functional/integration test but wasn't sure.

Checking a stubbed request context/headers should be simple enough to add. I'll update in the morning. I'll also add a changelog unless you don't think it necessary?

File: tests/functional/test_s3.py
Comment (nateprewitt) [inline tests/functional/test_s3.py:2178]: We can probably clean this up by using existing functions instead of rewriting the client creation function.
```suggestion
    op_kwargs = {
        "Bucket": "mybucket",
        "Key": "mykey",
        "Body": b"foo",
        "ChecksumAlgorithm": "sha256",
    }
    if input_param is not None:
        op_kwargs["ContentEncoding"] = content_encoding
    s3 = _create_s3_client(
        region='us-west-2', is_secure=True, s3_config={}, signature_version='s3v4'
    )
    with ClientHTTPStubber(s3) as http_stubber:
        http_stubber.add_response()
        s3.put_object(**op_kwargs)
        request_headers =  http_stubber.requests[-1].headers
        assert request_headers["Content-Encoding"] == expected_header
```

File: tests/functional/test_s3.py
Comment (nateprewitt) [inline tests/functional/test_s3.py:2152]:
```suggestion
    "content_encoding, expected_header",
```

File: tests/functional/test_s3.py
Comment (dlm6693) [reply tests/functional/test_s3.py:2178]: I thought about this but opted for consistency with the other checksum tests right above this especially since the test has nothing to with a secure request, custom config or signature version. I don't feel strongly about this though.

Comment (nateprewitt) [reply tests/functional/test_s3.py:2178]: Yeah, I think the existing checksum test case is the outlier rather than the rule. All other tests use `BaseS3OperationTest` or `_create_s3_client`. We should avoid continuing a bad pattern.

The args aren't super important but we can add sensible defaults to the `_create_s3_function` if needed like we're doing here. I'd say save that for a separate PR though to avoid conflating the two.

Comment (dlm6693) [reply tests/functional/test_s3.py]: Yeah totally I agree.

**Comment by dlm6693** [comment]
@nateprewitt I'll wait until #2860 is merged, then I'll update the functional test to rely on the defaults in the client constructor.

File: tests/functional/test_s3.py
Comment (nateprewitt) [inline tests/functional/test_s3.py:2164]: We probably still want to clean this up. It's going to be difficult to understand the failure because we've put multiple failure cases onto a single line.

```suggestion
        request_headers = http_stubber.requests[-1].headers
        assert request_headers["Content-Encoding"] == expected_header
```

Comment (nateprewitt) [reply tests/functional/test_s3.py:2164]: And just in case it's missed, we need the expected_header to be bytes.

Comment (dlm6693) [reply tests/functional/test_s3.py:2164]: Ok I'll shoot that up now.

**Review by nateprewitt — APPROVED** [review]
:shipit: Looks good once the tests wrap

**Pull request merged**
